pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,225 @@
+		if nodeName == "" {
+			// Wait for the PVC to get a node name before continuing
+			return false, waitForFirstConsumer, nil
+		}
+	}
+
+	return true, waitForFirstConsumer, nil
+}
+
+type pvcModifierFunc func(pvc *corev1.PersistentVolumeClaim, source client.Object)
+
+func (r *ReconcilerBase) createPVCPrime(pvc *corev1.PersistentVolumeClaim, source client.Object, waitForFirstConsumer bool, pvcModifier pvcModifierFunc) (*corev1.PersistentVolumeClaim, error) {
+	labels := map[string]string{
+		common.CDILabelKey: common.CDILabelValue,
+	}
+	annotations := make(map[string]string)
+	annotations[cc.AnnImmediateBinding] = ""
+	if waitForFirstConsumer {
+		annotations[AnnSelectedNode] = pvc.Annotations[AnnSelectedNode]
+	}
+	pvcPrimeName := PVCPrimeName(pvc)
+	pvcPrime := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        pvcPrimeName,
+			Namespace:   pvc.Namespace,
+			Labels:      labels,
+			Annotations: annotations,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes:      pvc.Spec.AccessModes,
+			Resources:        pvc.Spec.Resources,
+			StorageClassName: pvc.Spec.StorageClassName,
+			VolumeMode:       pvc.Spec.VolumeMode,
+		},
+	}
+	pvcPrime.OwnerReferences = []metav1.OwnerReference{
+		*metav1.NewControllerRef(pvc, schema.GroupVersionKind{
+			Group:   "",
+			Version: "v1",
+			Kind:    "PersistentVolumeClaim",
+		}),
+	}
+	util.SetRecommendedLabels(pvcPrime, r.installerLabels, "cdi-controller")
+
+	if !isPVCOwnedByDataVolume(pvc) {
+		requestedSize, _ := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+		// disk or image size, inflate it with overhead
+		requestedSize, err := cc.InflateSizeWithOverhead(r.client, requestedSize.Value(), &pvc.Spec)
+		if err != nil {
+			return nil, err
+		}
+		pvcPrime.Spec.Resources.Requests[corev1.ResourceStorage] = requestedSize
+
+	}
+	if pvcModifier != nil {
+		pvcModifier(pvcPrime, source)
+	}
+	if err := r.client.Create(context.TODO(), pvcPrime); err != nil {
+		return nil, err
+	}
+
+	r.recorder.Eventf(pvc, corev1.EventTypeNormal, createdPVCPrimeSucceesfully, createdPVCPrimeSucceesfully)
+	return pvcPrime, nil
+}
+
+func (r *ReconcilerBase) rebindPV(targetPVC, pvcPrime *corev1.PersistentVolumeClaim) error {
+	pv, err := r.getPV(pvcPrime.Spec.VolumeName)
+	if pv == nil {
+		return err
+	}
+
+	// Examine the claimref for the PV and see if it's still bound to PVC'
+	claimRef := pv.Spec.ClaimRef
+	if claimRef.Name != pvcPrime.Name || claimRef.Namespace != pvcPrime.Namespace || claimRef.UID != pvcPrime.UID {
+		return nil
+	}
+
+	// Rebind PVC to target PVC
+	pv.Annotations = make(map[string]string)
+	pv.Spec.ClaimRef = &corev1.ObjectReference{
+		Namespace:       targetPVC.Namespace,
+		Name:            targetPVC.Name,
+		UID:             targetPVC.UID,
+		ResourceVersion: targetPVC.ResourceVersion,
+	}
+	r.log.V(3).Info("Rebinding PV to target PVC", "PVC", targetPVC.Name)
+	if err := r.client.Update(context.TODO(), pv); err != nil {
```

Comment:
Why don't we delete `pvcPrime` before we update the `PV` `ClaimRef` to `targetPVC`? are we sure it's not racy and PVC controller won't override it as `pvcPrime.Spec.VolumeName` still points the `PV`?

Reply:
the volume name is updated according to the PV claimRef and not the other way around. Also if you delete the pvcPrime while it has the claimRef wouldn't the dynamically provisioned PV be deleted?
And also its whats done in k8s -> https://github.com/kubernetes-csi/lib-volume-populator/blob/master/populator-machinery/controller.go#L648-L673 :)

Reply:
Cool, I learned something new. I simply played with this flow [here](https://github.com/kubevirt/containerized-data-importer/pull/2683/commits/2e23d09e3cb6d0a76f406247d07339a28342ef2a#diff-f492548a4d272a2e8c736f701a769c82d4143960f57bd0dd30f1541b08d3ac79R1810-R1835). I guess it was different in my case as I used `PersistentVolumeReclaimRetain`? I saw that if I don't delete the original PVC it will take the PV again.

Reply:
It is said in the introduction that pvc prime will be deleted, but in the actual scenario, pvc prime still exists after the upload is completed. And no pv rebind phenomenon is observed. I don't understand this, can you explain it?